frontend/src/Routes/DynamicRoute/DynamicRoute.tsx

```diff
@@ -15,38 +15,24 @@ type DynamicRouteProps = {
   location?: Location;
 };
 
-type RoutePage = {
-  path: string;
-  exact?: boolean;
-};
-
-const checkPath = (pathname, { path, exact }: RoutePage) => {
-  const [, section] = pathname.split('/');
-  return matchPath(pathname, { path, exact }) || matchPath(pathname, { path: `/${section}${path}`, exact });
-};
-const DynamicRoute: React.FC<DynamicRouteProps> = ({ location }) => {
-  const [Component, setComponent] = React.useState<React.ExoticComponent<any>>(React.Fragment);
-  const [currClassName, setCurrClassName] = React.useState<string>(null);
-  // TODO: This seems to get executed when the tree fails to have PluginStoreProvider and then explodes
+const DynamicRoute: React.FC<DynamicRouteProps> = () => {
   const dynamicRoutePages = useExtensions<DynamicRoutePage>(isDynamicRoutePage);
-  React.useEffect(() => {
-    if (location) {
-      const { properties: currRoute, pluginName } =
-        dynamicRoutePages.find(({ properties }) => {
-          if (Array.isArray(properties.path)) {
-            return properties.path.some((path) => checkPath(location.pathname, { ...properties, path }));
-          }
-          return checkPath(location.pathname, properties as RoutePage);
-        }) || {};
-      if (currRoute) {
-        setCurrClassName(camelCase(pluginName));
-        setComponent(() =>
-          React.lazy(async () => {
+
+  const routes = React.useMemo(
+    () =>
+      dynamicRoutePages.map(({ properties: { component, ...currRoute }, pluginName, uid }) => {
+        return {
+          ...currRoute,
+          uid,
+          className: camelCase(pluginName),
+          Component: React.lazy(async () => {
```

Comment:
@karelhala @vojtechszocs 
if the `dynamicRoutePages` changes, will the use of `React.lazy` here create a new component which causes the old one to get unmounted an re-rendered? If so, we may need to memoize this value.

Reply:
@christianvogt @karelhala

> if the `dynamicRoutePages` changes, will the use of `React.lazy` here create a new component

Yes. The `Component` property above will reference a new lazy component instance.

The `component()` Promise will, however, resolve to the same "actual" React component instance as provided by dynamic plugins.

> which causes the old one to get unmounted

Yes, because the same "actual" React component is wrapped in a newly created lazy component.

> If so, we may need to memoize this value.

The "actual" React components are accessed via Promise so we can't depend on them directly since that would be an async dependency.

In OpenShift plugin SDK, the `PluginStore` guarantees that all `a.b` code references (for the given plugin) point to the same `CodeRef` instance - see `codeRefKey` and `codeRefCache` usage in [coderefs.ts](https://github.com/openshift/dynamic-plugin-sdk/blob/main/packages/lib-core/src/store/coderefs.ts).

Therefore, we could memoize lazy components based on `CodeRef` references to the "actual" components.

However, I think this would involve the use of React refs since we want to remember component `CodeRef`s across renders, instead of always deriving that information from `dynamicRoutePages` (as is the usual pattern with React hooks, e.g. get A from hookA, get B based on A from hookB etc).

**TL;DR** if this code works, I'd keep it and not worry about the extra re-renders unless we are really concerned about the performance. But ideally, yes - we should memoize all lazy component instances that defer to `CodeRef` components, assuming referential stability of `CodeRef`s. Maybe we should introduce a utility hook for this in OpenShift plugin SDK. cc @florkbr

Reply:
Good point! I'd keep it as is as well, we can track if changes to `dynamicRoutePages` causes a lot of re-renders, if it'll be the case we could optimize later. After all we want to trigger a re-render when this prop changes and react-router will do the heavy lifting of preventing the extra renders.

Reply:
This would only affect an env with several plugins that load in asynchronously which add their own routes.